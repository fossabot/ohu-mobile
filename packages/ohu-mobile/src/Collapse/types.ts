import { IconDef } from '../types';

export type CollapseExpandIconPosition = 'left' | 'right';

export type CollapseValue = (string | number)[] | string | number;

export interface CollapseProps {
  expandIcon?: IconDef | null;
  expandIconPosition?: CollapseExpandIconPosition;
  value?: CollapseValue;
  accordion?: boolean;
}

export interface CollapseEvents {
  onChange: CollapseValue;
}

export interface CollapseItemProps {
  key: string | number;
  title?: string;
  hasList?: boolean;
  disabled?: boolean;
  loading?: boolean;
}

export interface CollapseItemEvents {
  onExpand: string | number;
  onExpanded: void;
  onShrink: string | number;
  onShrinked: void;
}
