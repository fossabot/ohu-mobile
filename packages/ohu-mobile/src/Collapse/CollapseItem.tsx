import { defineDescendantComponent, props } from '../_utils/defineComponent';
import { CollapseItemProps, CollapseItemEvents } from './types';
import Collapse from './Collapse';
import Icon from '../Icon';
import Divider from '../Divider';
import animate from '../_utils/animate';
import { easeOutQuint } from '../_utils/easing';

interface CollapseItemMethods {
  getExpandState: () => boolean;
}

export default defineDescendantComponent<InstanceType<typeof Collapse>, CollapseItemProps, CollapseItemEvents, {}, CollapseItemMethods>(
  'collapse',
  'collapse-item'
).create({
  props: {
    title: props(String).default(''),
    hasList: props(Boolean).default(false),
    disabled: props(Boolean).default(false),
  },
  computed: {
    expandState() {
      return this.getExpandState();
    },
  },
  data() {
    return {
      startHeight: -1,
      contentHeight: -1,
    };
  },
  methods: {
    getExpandState() {
      if (this.disabled) return false;
      const key = this.$vnode.key;
      const value = this.ancestor.state;
      if (!key || !value) return false;
      return value instanceof Array ? value.indexOf(key) >= 0 : value === key;
    },
    handleHeaderClick(expandState: boolean) {
      if (this.disabled) return;
      const key = this.$vnode.key;
      if (key) {
        this.ancestor.itemChange(key, !expandState);
      }
      if (this.expandState) {
        this.$emit('expand');
      } else {
        this.$emit('shrink');
      }
    },
    beforeContentEnter(el: HTMLElement) {
      el.style.height = 'auto';
      this.$nextTick(() => {
        animate(el, {
          timingFunction: easeOutQuint,
          duration: 300,
          done() {
            el.style.height = '';
          },
          props: {
            height: [0, el.offsetHeight],
          },
        });
      });
    },
    contentLeave(el: HTMLElement, done: () => void) {
      el.style.height = `${el.offsetHeight}px`;
      animate(el, {
        timingFunction: easeOutQuint,
        duration: 300,
        done: () => {
          el.style.height = '';
          done();
        },
        props: {
          height: 0,
        },
      });
    },
  },
  render() {
    const { $slots, title, ancestor, expandState, hasList, disabled } = this;
    const root = this.root();
    const header = root.element('header');
    const content = root.element('content');
    const contentStyle: Partial<CSSStyleDeclaration> = {};
    if (!expandState) {
      contentStyle.display = 'none';
      contentStyle.height = '0';
    }
    return (
      <div class={root.is([expandState && 'expand', disabled && 'disabled' ])}>
        <div
          class={header.is([ancestor.expandIcon ? ancestor.expandIconPosition : ''])}
          role="button"
          tabindex={-1}
          aria-expanded={expandState}
          onClick={() => this.handleHeaderClick(expandState)}>
          <div class={header.element('text')}>
            {$slots.title || title}
          </div>
          {
            ancestor.expandIcon
            &&
            <div class={header.element('icon')}>
              <Icon type={ancestor.expandIcon}></Icon>
            </div>
          }
        </div>
        <transition
          css={false}
          onBeforeEnter={(el) => this.beforeContentEnter(el as HTMLElement)}
          onLeave={(el, done) => this.contentLeave(el as HTMLElement, done)}>
          <div class={content} v-show={expandState}>
            <Divider />
            <div class={content.element('inner').has([ hasList && 'list' ])}>{$slots.default}</div>
          </div>
        </transition>
      </div>
    );
  },
});
