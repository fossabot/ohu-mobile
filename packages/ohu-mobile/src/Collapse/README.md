# Collapse

## Usage

```js
import { Collapse } from 'ohu-mobile';
```

## Design Goals

+ Supoort two modes: expand single area at a time; multiple areas can be expanded at the same time.

+ `expandIcon` prop (can be `null`)

+ `expandIconPosition` prop(`left` or `right`)

+ CollapseItem can be disabled
