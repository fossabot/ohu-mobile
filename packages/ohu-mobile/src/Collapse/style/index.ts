import '../../_styles';
import './index.scss';

// style dependencies
import '../../Divider/style';
import '../../Icon/style';
