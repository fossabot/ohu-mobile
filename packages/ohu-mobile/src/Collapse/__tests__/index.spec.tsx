import { mount } from '@vue/test-utils';
import Vue from 'vue';
import Collapse from '..';


/**
 * 组件单元测试要点：
 * 1. 保证DOM结构正确
 * 2. 事件触发正确
 */

describe('Collapse', () => {
  it('render correct', () => {
    const Component = Vue.extend({
      render() {
        return (
          <Collapse value={1}>
            <Collapse.Item key={1}></Collapse.Item>
          </Collapse>
        );
      },
    });
    const wrapper = mount(Component);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should remove expandIcon', () => {
    const Component = Vue.extend({
      render() {
        return (
          <Collapse expandIcon={null}>
            <Collapse.Item title={'should remove expandIcon'} key={1}></Collapse.Item>
          </Collapse>
        );
      },
    });
    const wrapper = mount(Component);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('v-model value', () => {
    const Component = Vue.extend({
      data() {
        return {
          active: [1],
        };
      },
      render() {
        return (
          <Collapse v-model={this.active}>
            <Collapse.Item title="should open panel" key={1}></Collapse.Item>
            <Collapse.Item title="should open panel" key={2}></Collapse.Item>
            <Collapse.Item title="should open panel" key={3}></Collapse.Item>
          </Collapse>
        );
      },
    });
    const wrapper = mount(Component);
    wrapper.setData({ active: [2] });
    setTimeout(() => {
      expect(wrapper.findAll('.ohu-collapse-item').at(1).contains('.is-expand')).toBe(true);
    }, 0);
  });

  it('v-model `change` event', () => {
    const func = jest.fn();
    const Component = Vue.extend({
      render() {
        return (
          <Collapse onChange={func}>
            <Collapse.Item title="1" key="1"></Collapse.Item>
            <Collapse.Item title="2" key="2"></Collapse.Item>
          </Collapse>
        );
      },
    });
    const wrapper = mount(Component);
    wrapper.find('.ohu-collapse-item__header').trigger('click');
    expect(func).toBeCalled();
  });

  it('should open panel', () => {
    const Component = Vue.extend({
      data() {
        return {
          active: [1],
        };
      },
      render() {
        return (
          <Collapse v-model={this.active}>
            <Collapse.Item title="should open panel" key={1}></Collapse.Item>
            <Collapse.Item title="should open panel" key={2}></Collapse.Item>
            <Collapse.Item title="should open panel" key={3}></Collapse.Item>
          </Collapse>
        );
      },
    });
    const wrapper = mount(Component);
    const items = wrapper.findAll('.ohu-collapse-item__header');
    items.at(0).trigger('click');
    items.at(1).trigger('click');
    items.at(2).trigger('click');
    setTimeout(() => {
      expect(wrapper.vm.active).toBe([1, 2, 3]);
    }, 300);
  });

  it('should open one panel in `accordion` mode', () => {
    const Component = Vue.extend({
      render() {
        return (
          <Collapse accordion value={1}>
            <Collapse.Item title="1" key={1}>1</Collapse.Item>
            <Collapse.Item title="2" key={2}>2</Collapse.Item>
            <Collapse.Item title="3" key={3}>3</Collapse.Item>
          </Collapse>
        );
      },
    });
    const wrapper = mount(Component);
    const items = wrapper.findAll('.ohu-collapse-item');
    expect(items.length).toBe(3);
    if (items.length > 0) {
      expect(items.at(0).contains('.is-expand')).toBe(true);
      wrapper.findAll('.ohu-collapse-item__header').at(1).trigger('click');
      setTimeout(() => {
        expect(items.at(0).contains('.is-expand')).toBe(false);
        expect(items.at(1).contains('.is-expand')).toBe(true);
      }, 300);
    }
  });
});
