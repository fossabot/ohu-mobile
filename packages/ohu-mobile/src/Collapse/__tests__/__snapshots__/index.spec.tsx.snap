// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse render correct 1`] = `
<div
  class="ohu-collapse"
>
  <div
    class="ohu-collapse-item is-expand"
  >
    <div
      aria-expanded="true"
      class="ohu-collapse-item__header is-right"
      role="button"
      tabindex="-1"
    >
      <div
        class="ohu-collapse-item__header__text"
      />
      <div
        class="ohu-collapse-item__header__icon"
      >
        <svg
          aria-hidden="true"
          class="ohu-icon"
          focusable="false"
          role="presentation"
          tabindex="-1"
          title="arrow-down-s"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 13.172l4.95-4.95 1.414 1.414L12 16 5.636 9.636 7.05 8.222z"
          />
        </svg>
      </div>
    </div>
    <div
      class="ohu-collapse-item__content"
    >
      <div
        class="ohu-divider is-horizontal"
        role="separator"
      />
      <div
        class="ohu-collapse-item__content__inner"
      />
    </div>
  </div>
</div>
`;

exports[`Collapse should remove expandIcon 1`] = `
<div
  class="ohu-collapse"
>
  <div
    class="ohu-collapse-item"
  >
    <div
      class="ohu-collapse-item__header is-"
      role="button"
      tabindex="-1"
    >
      <div
        class="ohu-collapse-item__header__text"
      >
        should remove expandIcon
      </div>
    </div>
    <div
      class="ohu-collapse-item__content"
      style="display: none;"
    >
      <div
        class="ohu-divider is-horizontal"
        role="separator"
      />
      <div
        class="ohu-collapse-item__content__inner"
      />
    </div>
  </div>
</div>
`;
